### PR TITLE
feat($scaffold): composable plugins

### PR DESCRIPTION
**Pull request by ghost**
Example changes for sample plugin generator enhancements

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -15,6 +15,9 @@ cache:
   - composer
   - $HOME/.composer/cache
 
+install:
+  - composer install --no-interaction
```
Comment (ghost) [inline .travis.yml:19]: Start moving away from `global` installs

File: .travis.yml [hunk 2]
```diff
@@ -31,6 +34,9 @@ matrix:
       env: WP_TRAVISCI=phpcs
     - php: 5.3
       env: WP_VERSION=latest
+  allow_failures:
+    - php: 5.6
+      env: WP_TRAVISCI=phpcs
```
Comment (ghost) [inline .travis.yml:39]: Allow failures in PHPCS for standard plugins to eliminate churn for common use cases.

File: .travis.yml [hunk 3]
```diff
@@ -46,7 +52,7 @@ before_script:
   - |
     if [[ "$WP_TRAVISCI" == "phpcs" ]] ; then
       composer global require wp-coding-standards/wpcs
-      phpcs --config-set installed_paths $HOME/.composer/vendor/wp-coding-standards/wpcs
+      composer run lint:php:config
```
Comment (ghost) [inline .travis.yml:55]: Enables use of commands via composer scripts from development environment.

File: composer.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        "lint:php:config": "phpcs --config-set installed_paths vendor/wp-coding-standards/wpcs",
+        "lint:php": "phpcs --standard=phpcs.ruleset.xml $(find src tests -name '*.php')",
+        "lint:php:fix": "phpcs --standard=phpcs.ruleset.xml $(find src tests -name '*.php')"
+    },
+    "require": {
+        "php": ">=5.3",
+        "composer/installers": "^1.3",
+        "wp-coding-standards/wpcs": "*"
+    }
+}
```
Comment (ghost) [inline composer.json:15]: Purposely omitted `license` field. Includes basic setup for `type="wordpress-plugin"` for easy plugin installation via composer. Adds scripts to configure, lint and fix PHP code. Commands are namespaced to allow planned introduction of `lint:js:*` scripts.

**Comment by ghost** [comment]
Code annotated. I'd like to discuss changes here for testing purposes before subbing a pull to the generator repo. Your review is appreciated.

File: .travis.yml (hunk 1 above)
Comment (danielbachhuber) [reply .travis.yml:19]: Why?

Comment (ghost) [reply .travis.yml:19]: Off the cuff, two reasons: security and parity with development environment. I was surprised to see global installs were even allowed with `sudo: false` TBH. But that I'm guessing is a benefit of Travis' container infrastructure.

**Comment by danielbachhuber** [comment]
@jhabdas Generally, I'm on the fence about including these changes. They seem fine (e.g. they probably work), but they're opinionated changes and I'm not sure about imposing these opinions. If something already works, there has to be a strongly compelling reason to change it.

There's also some historical context for why things are the way they are. Notably, `phpcs` is added via `wp scaffold plugin-tests`, which doesn't assume the presence of a `composer.json` file. See https://github.com/wp-cli/wp-cli/issues/1816 and https://github.com/wp-cli/wp-cli/issues/310

**Comment by ghost** [comment]
> If something already works, there has to be a strongly compelling reason to change it.

No offense meant, but that's not the first time I've noticed that motto from the WP team.

The compelling reason to make these changes is to advance the plugin architecture and prevent individuals from reinventing the wheel. From my standpoint, if this is a sample plugin generated by CLI it should be advancing constantly and help push the `wpcs` forward, which is an immediate benefit of these changes.

**EDIT:** Re: `wp scaffold plugin-tests` command. Didn't work for me by the time I actually discovered them via Ben Balter's repo after searching for a good way to build a plug-in for two weeks.

**Comment by danielbachhuber** [comment]
> No offense meant, but that's not the first time I've noticed that motto from the WP team, e.g.

It seems like you're trying to start debates with as many aspects of the WordPress project that you can. I'm not really interested in having a debate here.

> The compelling reason to make these changes is to advance the plugin architecture and prevent individuals from reinventing the wheel. From my standpoint, if this is a sample plugin generated by CLI it should be advancing constantly and help push the `wpcs` forward, which is an immediate benefit of these changes.

Sure, that's an opinionated reason though, not a pragmatic, practical one.

I'm open to considering each change from a pragmatic, practical perspective. I'm not interested in debating opinions.

**Comment by ghost** [comment]
I don't want to get off topic. Let's please take the changes at their face value. But I did want to raise a clear frustration with what I perceive to be a lack of embracement of change in the organization.

File: .gitignore [hunk 5]
```diff
@@ -5,3 +5,5 @@ node_modules/
 *.sql
 *.tar.gz
 *.zip
+
+/vendor/
```
Comment (danielbachhuber) [inline .gitignore:9]: This seems fine

File: .travis.yml (hunk 1 above)
Comment (danielbachhuber) [reply .travis.yml:19]: I don't want to change the install step.

File: .travis.yml (hunk 2 above)
Comment (danielbachhuber) [reply .travis.yml:39]: If you want to allow failures for `phpcs`, you can do so by modifying your `.travis.yml` after scaffolding. I'd like to keep failures as the default behavior, to encourage following WPCS

**Comment by ghost** [comment]
> Sure, that's an opinionated reason though, not a pragmatic, practical one.

FWIW - I'm sensing opinion coming through. I'm not clear how one might justify these changes are not pragmatic. They reek of pragmatism, and are opinionated insomuch as the entire PHP world resolves around a single package manager called Composer.

File: .travis.yml (hunk 3 above)
Comment (danielbachhuber) [reply .travis.yml:55]: This would require programmatically modifying `composer.json`, which I don't want to do.

**Comment by ghost** [comment]
@danielbachhuber would you mind starting a formal review. I'd like to have two captains sign off or punt on this if possible, so I can walk away knowing how to feel.

**Comment by danielbachhuber** [comment]
> would you mind starting a formal review.

I've left some comments.

> I'd like to have two captains sign off or punt on this if possible, so I can walk away knowing how to feel.

I'm not sure what you mean by this?

**Comment by ghost** [comment]
> I've left some comments.

A formal review in GitHub does not poke and prod. It initiates a process via the review button, allowing a sign-off or clear declination of changes so they don't get lost in the shuffle of comments or any kind of bike shedding

> I'm not sure what you mean by this?

If you absorb the issues I linked above the clear theme is a distributed command and control structure without peer review, which is what what a two-review system helps avoid in addition to mitigating bus factor.

**Comment by danielbachhuber** [comment]
@jhabdas Please take your contributions elsewhere, thanks.

**Comment by jhabdas** [comment]
> Please take your contributions elsewhere, thanks.

Whether its pride standing in in the way or a growth opportunity in accepting candid feedback I cannot say, but this response was from a place of emotion and serves only to prove my points made above regarding command and control and getting off topic. The two safest options I can see is tools like this are either moved into or a contribution policy added so it's not outpaced by the next gunslinger who comes by. But I digress. Thanks for looking. And I hope you have a chance to reflect on your behavior.

**Pull request closed without merging**